<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Изучение-общей-информации-о-данных-в-таблице" data-toc-modified-id="Изучение-общей-информации-о-данных-в-таблице-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Изучение общей информации о данных в таблице</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Удаление-пропусков" data-toc-modified-id="Удаление-пропусков-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Удаление пропусков</a></span></li><li><span><a href="#Обработка-аномальных-значений" data-toc-modified-id="Обработка-аномальных-значений-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Обработка аномальных значений</a></span></li><li><span><a href="#Удаление-пропусков-(продолжение)" data-toc-modified-id="Удаление-пропусков-(продолжение)-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Удаление пропусков (продолжение)</a></span></li><li><span><a href="#Изменение-типов-данных" data-toc-modified-id="Изменение-типов-данных-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Изменение типов данных</a></span></li><li><span><a href="#Обработка-дубликатов" data-toc-modified-id="Обработка-дубликатов-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Обработка дубликатов</a></span></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Категоризация данных</a></span></li></ul></li><li><span><a href="#Исследование-данных-и-ответы-на-вопросы" data-toc-modified-id="Исследование-данных-и-ответы-на-вопросы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Исследование данных и ответы на вопросы</a></span><ul class="toc-item"><li><span><a href="#Существует-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Существует-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Существует ли зависимость между количеством детей и возвратом кредита в срок?</a></span></li><li><span><a href="#Существует-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Существует-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Существует ли зависимость между семейным положением и возвратом кредита в срок?</a></span></li><li><span><a href="#Существует-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Существует-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Существует ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span></li><li><span><a href="#Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Как разные цели кредита влияют на его возврат в срок?</a></span></li><li><span><a href="#Привести-возможные-причины-появления-пропусков-в-исходных-данных." data-toc-modified-id="Привести-возможные-причины-появления-пропусков-в-исходных-данных.-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Привести возможные причины появления пропусков в исходных данных.</a></span></li><li><span><a href="#Почему-заполнить-пропуски-медианным-значением-—-лучшее-решение-для-количественных-переменных?" data-toc-modified-id="Почему-заполнить-пропуски-медианным-значением-—-лучшее-решение-для-количественных-переменных?-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Почему заполнить пропуски медианным значением — лучшее решение для количественных переменных?</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

# Проект: исследование надёжности заёмщиков <a class="tocSkip">

Цель проекта - разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок

## Изучение общей информации о данных в таблице

Импортируем библиотеку pandas, считаем данные из csv-файла в датафрейм и сохраним в переменную `data`

In [1]:
import pandas as pd
data = pd.read_csv('data.csv')

Выведем первые 20 строк датафрейма на экран

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

Выведем основную информацию о датафрейме

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


В столбцах days-employed и total_income обнаружены пропуски

## Предобработка данных

### Удаление пропусков

Выведем количество пропущенных значений для каждого столбца

In [4]:
data.isnull().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Обработаем значения в столбце total_income (хранит данные о доходах). На сумму дохода сильнее всего влияет тип занятости, поэтому в этом столбце заполняем медианным значением по каждому типу из столбца income_type

In [5]:
for type in data['income_type'].unique():
    data.loc[(data['income_type'] == type) & (data['total_income'].isna()), 'total_income'] = data.loc[(data['income_type'] == type), 'total_income'].median()
# проверим на заполнение пропусков
print(data['total_income'].isna().sum())

0


### Обработка аномальных значений

Заменим отрицательные значения в столбце days_employed на положительные

In [6]:
data['days_employed'] = data['days_employed'].abs()

Для каждого типа занятости выведем медианное значение трудового стажа days_employed в днях

In [7]:
data.groupby('income_type')['days_employed'].median()

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) видим аномально большие значения. Исправить такие значения сложно, поэтому оставляем как есть

Выведем перечень уникальных значений столбца children

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

Видим два аномальных значения. Удалим строки, в которых они встречаются

In [9]:
data = data.loc[(data['children'] != -1) & (data['children'] != 20)]

Выведем перечень уникальных значений столбца children снова, чтобы убедиться, что артефакты удалены

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

Заполним пропуски в столбце days_employed медианными значениями по каждому типу занятости income_type

In [11]:
for type in data['income_type'].unique():
    data.loc[(data['income_type'] == type) & (data['days_employed'].isna()), 'days_employed'] = data.loc[(data['income_type'] == type), 'days_employed'].median()

Убедимся, что все пропуски заполнены

In [12]:
data['days_employed'].isna().sum()

0

### Изменение типов данных

Заменим вещественный тип данных в столбце total_income на целочисленный

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

Обработаем неявные дубликаты в столбце education, приведём их к нижнему регистру

In [14]:
data['education'] = data['education'].str.lower()

Выведем на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалим их

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

На основании диапазонов, указанных ниже, создадим в датафрейме столбец total_income_category с категориями:

0–30000 — 'E';
30001–50000 — 'D';
50001–200000 — 'C';
200001–1000000 — 'B';
1000001 и выше — 'A'.

Используем собственную функцию categorize_income() и метод apply()

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income > 1000000:
            return 'A'
    except:
        'значение вне диапазона'

Применим функцию

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

Выведем на экран перечень уникальных целей взятия кредита из столбца purpose

In [19]:
print(data['purpose'].unique())

['покупка жилья' 'приобретение автомобиля' 'дополнительное образование'
 'сыграть свадьбу' 'операции с жильем' 'образование'
 'на проведение свадьбы' 'покупка жилья для семьи' 'покупка недвижимости'
 'покупка коммерческой недвижимости' 'покупка жилой недвижимости'
 'строительство собственной недвижимости' 'недвижимость'
 'строительство недвижимости' 'на покупку подержанного автомобиля'
 'на покупку своего автомобиля' 'операции с коммерческой недвижимостью'
 'строительство жилой недвижимости' 'жилье'
 'операции со своей недвижимостью' 'автомобили' 'заняться образованием'
 'сделка с подержанным автомобилем' 'получение образования' 'автомобиль'
 'свадьба' 'получение дополнительного образования' 'покупка своего жилья'
 'операции с недвижимостью' 'получение высшего образования'
 'свой автомобиль' 'сделка с автомобилем' 'профильное образование'
 'высшее образование' 'покупка жилья для сдачи' 'на покупку автомобиля'
 'ремонт жилью' 'заняться высшим образованием']


Создадим функцию, которая на основании данных из столбца purpose сформирует новый столбец purpose_category, в который войдут следующие категории:

'операции с автомобилем',
'операции с недвижимостью',
'проведение свадьбы',
'получение образования'.
Например, если в столбце purpose находится подстрока 'на покупку автомобиля', то в столбце purpose_category должна появиться строка 'операции с автомобилем'.

Используем собственную функцию с именем categorize_purpose() и метод apply()

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'недвиж' in row or 'жил' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образ' in row:
            return 'получение образования'
    except:
        return 'категория отсутствует'

Применим функцию

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

In [22]:
print(data.head(20))

    children  days_employed  dob_years            education  education_id  \
0          1    8437.673028         42               высшее             0   
1          1    4024.803754         36              среднее             1   
2          0    5623.422610         33              среднее             1   
3          3    4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0     926.185831         27               высшее             0   
6          0    2879.202052         43               высшее             0   
7          0     152.779569         50              среднее             1   
8          2    6929.865299         35               высшее             0   
9          0    2188.756445         41              среднее             1   
10         2    4171.483647         36               высшее             0   
11         0     792.701887         40              среднее             1   

## Исследование данных и ответы на вопросы

### Существует ли зависимость между количеством детей и возвратом кредита в срок?

Построим сводную таблицу со столбцами: children, count, sum, debt_ratio, где считаем по столбцу debt, а столбец debt_ratio - процент просроченных платежей для каждой категории клиентов

In [23]:
data_pivot = data.pivot_table(index=['children'], values='debt', aggfunc={'count','sum'})
data_pivot['debt_ratio'] = round((data_pivot['sum'] / data_pivot['count']) * 100, 2)
data_pivot = data_pivot.sort_values(by='debt_ratio', ascending=True)
display(data_pivot)

count   sum  debt_ratio
children                         
5             9     0        0.00
0         14091  1063        7.54
3           330    27        8.18
1          4808   444        9.23
2          2052   194        9.45
4            41     4        9.76

Вывод: зависимость между количеством детей и возвратом кредита в срок прослеживается не сильно, но она есть. В семьях с 5-ю детьми задолженностей нет. Вероятно, из-за того, что такие семьи вполне обеспечены. Клиенты без детей имеют чуть меньшее количество долгов по сравнению с семьями, где 1-2-3 ребёнка. 4 ребёнка - серьёзная нагрузка на семью, здесь почти в каждом 10-м случае просрочка по платежам

### Существует ли зависимость между семейным положением и возвратом кредита в срок?

Аналогично, построим сводную таблицу со столбцами: family_status, count, sum, debt_ratio, где считаем по столбцу debt, а столбец debt_ratio - процент просроченных платежей для каждой категории клиентов

In [24]:
data_pivot = data.pivot_table(index=['family_status'], values='debt', aggfunc={'count','sum'})
data_pivot['debt_ratio'] = round((data_pivot['sum'] / data_pivot['count']) * 100, 2)
data_pivot = data_pivot.sort_values(by='debt_ratio', ascending=True)
display(data_pivot)

count  sum  debt_ratio
family_status                                
вдовец / вдова           951   63        6.62
в разводе               1189   84        7.06
женат / замужем        12261  927        7.56
гражданский брак        4134  385        9.31
Не женат / не замужем   2796  273        9.76

Вывод: зависимость прослеживается, но она незначительна, как и в предыдущем случае. Вдовцы/вдовы и те, кто был в браке, возвращают кредиты стабильнее, чем остальные категории. Скорее всего сказываются такие факторы как доход (не в браке обычно люди молодые и менее состоятельные) и вероятное наличие детей и других обязательств (дополнительная финансовая нагрузка на женатых).

### Существует ли зависимость между уровнем дохода и возвратом кредита в срок?

Схема та же, что в предыдущих построениях, меняем целевой столбец в сводной таблице: total_income_category, count, sum, debt_ratio. Категории составляли ранее: 0–30000 — 'E'; 30001–50000 — 'D'; 50001–200000 — 'C'; 200001–1000000 — 'B'; 1000001 и выше — 'A'

In [25]:
data_pivot = data.pivot_table(index=['total_income_category'], values='debt', aggfunc={'count','sum'})
data_pivot['debt_ratio'] = round((data_pivot['sum'] / data_pivot['count']) * 100, 2)
data_pivot = data_pivot.sort_values(by='debt_ratio', ascending=True)
display(data_pivot)

count   sum  debt_ratio
total_income_category                         
D                        349    21        6.02
B                       5014   354        7.06
A                         25     2        8.00
C                      15921  1353        8.50
E                         22     2        9.09

Вывод: здесь уже просматривается некоторая логика. Клиенты с самым низким доходом (категория Е) имеют наивысший процент невозврата по кредитам из-за того, что они часто живут "от получки до получки" и кредит сильно бьёт по и без того маленькому бюджету.  Категория С - самая массовая в выборке, довольно высокий процент невозврата кредита в срок можно объяснить семейным положением/наличем детей/кредитами в других банках и т.д. Примечательно, что клиенты с самым высоким доходом (категория А) - не самые надёжные плательщики.

### Как разные цели кредита влияют на его возврат в срок?

Меняем целевой столбец в сводной таблице: purpose_category, count, sum, debt_ratio

In [26]:
data_pivot = data.pivot_table(index=['purpose_category'], values='debt', aggfunc={'count','sum'})
data_pivot['debt_ratio'] = round((data_pivot['sum'] / data_pivot['count']) * 100, 2)
data_pivot = data_pivot.sort_values(by='debt_ratio', ascending=True)
display(data_pivot)

count  sum  debt_ratio
purpose_category                                
операции с недвижимостью  10751  780        7.26
проведение свадьбы         2313  183        7.91
получение образования      3988  369        9.25
операции с автомобилем     4279  400        9.35

Вывод: недвижимость - более "серьёзная" категория для взятия кредита. Здесь нужно брать на себя больше обязательств и, соответственно, выплачивать кредит в срок. Сумма тут тоже обычно больше, а следовательно, больше и процент за просрочку. Чего стараются избегать. "Образовательная" категория - в основном, для молодых людей, часто не обладающих стабильным доходом. Почему на вершине антирейтинга стоит "автомобильная" категория, по данной выборке объяснить трудно. Вероятно, нужно смотреть другие метрики

### Привести возможные причины появления пропусков в исходных данных.

Пропуски были в столбцах days_employed и total_income. Обычно это результат человеческого фактора, но возможно, существует зависимость в заполнении столбцов, на что косвенно указывает одинаковое количество пропусков в них (по 2174)

### Почему заполнить пропуски медианным значением — лучшее решение для количественных переменных?

Среднее арифметическое не учитывает выброс данных, медианное значение лучше отражает выборку, особенно для большого количества данных

## Общий вывод

Цель проекта состояла в том, чтобы выяснить, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Гипотезы подтвердить в полной мере не представляется возможным, но существуют некоторые зависимости. Например, семья с 2 и 4 детьми скорее всего просрочит выплату по кредиту. Как и состоящие в гражданском браке, а также неженатые/незамужние. Самые надёжные плательщики, согласно проверке: разведённые люди с доходом около 100 тыс., берущие кредит на операции с недвижимостью
